chessington/engine/pieces.py

```diff
@@ -4,85 +4,244 @@
-    def get_available_moves(self, board: Board) -> List[Square]:
-        """
+	@abstractmethod
+	def get_available_moves(self, board: Board) -> List[Square]:
+		"""
         Get all squares that the piece is allowed to move to.
         """
-        pass
+		pass
 
-    def move_to(self, board, new_square):
-        """
+	def move_to(self, board, new_square):
+		"""
         Move this piece to the given square on the board.
         """
-        current_square = board.find_piece(self)
-        board.move_piece(current_square, new_square)
+		current_square = board.find_piece(self)
+		board.move_piece(current_square, new_square)
+
+	def is_position_on_board(self, new_square: Square) -> bool:
+		return 0 <= new_square.row < BOARD_SIZE and 0 <= new_square.col < BOARD_SIZE
 
 
 class Pawn(Piece):
-    """
+	"""
     A class representing a chess pawn.
     """
-    def get_available_moves(self, board) -> List[Square]:
-        current_square = board.find_piece(self)
-        if self.player == Player.BLACK:
-            square_in_front = Square.at(current_square.row - 1, current_square.col)
-            return [square_in_front]
-        else:
-            square_in_front = Square.at(current_square.row + 1, current_square.col)
-            return [square_in_front]
+
+	def check_square_in_front(self, board, square: Square) -> bool:
+		return self.is_position_on_board(square) and board.get_piece(square) is None
+
+	def check_two_squares_in_front(self, board, row, square_in_front: Square, two_squares_in_front: Square) -> bool:
+		if self.player == Player.BLACK:
+			return (row == 6 and board.get_piece(square_in_front) is None
+					and board.get_piece(two_squares_in_front) is None)
+		else:
+			return (row == 1 and board.get_piece(square_in_front) is None
+					and board.get_piece(two_squares_in_front) is None)
+
+	def check_attack(self, board, square: Square) -> bool:
+		if self.is_position_on_board(square):
+			piece_attacked = board.get_piece(square)
+
+			if self.player == Player.BLACK:
+				return piece_attacked is not None and piece_attacked.player is Player.WHITE
+			else:
+				return piece_attacked is not None and piece_attacked.player is Player.BLACK
+
+		return False
+
+	def get_available_moves(self, board) -> List[Square]:
+		current_square = board.find_piece(self)
+		available_moves = []
+
+		if self.player == Player.BLACK:
+			square_in_front = Square.at(current_square.row - 1, current_square.col)
+			two_squares_in_front = Square.at(current_square.row - 2, current_square.col)
+			attack_left = Square.at(current_square.row - 1, current_square.col - 1)
+			attack_right = Square.at(current_square.row - 1, current_square.col + 1)
+
+		else:
+			square_in_front = Square.at(current_square.row + 1, current_square.col)
+			two_squares_in_front = Square.at(current_square.row + 2, current_square.col)
+			attack_left = Square.at(current_square.row + 1, current_square.col - 1)
+			attack_right = Square.at(current_square.row + 1, current_square.col + 1)
```

Comment:
This can be commonised considering that the the only difference is that `Black` moves the pawn at the same position as `White`, but `* (-1)`.